README.md
# `Read MoreX` [![pub package](https://img.shields.io/pub/v/read_morex.svg)](https://pub.dartlang.org/packages/read_morex)

The `read_morex` package is a Flutter widget that allows you to manage the display of long text with a "Read More" button that lets users expand to view the entire text or collapse it to a shorter version. It allows you to set a maximum number of lines and a maximum length for the text. When the content exceeds these limits, it provides a 'Read more' link, allowing users to see the full content, and a 'Show less' link to collapse it again. You can also apply custom text filters using regular expressions to highlight or modify specific patterns within the text. This widget offers flexibility and control over how text content is displayed and filtered within your Flutter app.

![Example Screenshot](https://github.com/tro1d/read_morex/blob/master/example/read_morex_example.gif?raw=true)

## Installation

You can import the `read_morex` package into your Flutter project by adding the following dependency to your `pubspec.yaml` file:
```yaml
dependencies:
  read_morex: ^1.1.0
```
Don't forget to run flutter pub get after adding this dependency.

## Usage
Here's how to use it:
#### Import the read_morex package:
```Dart
import 'package:read_morex/read_morex.dart';
```
#### Use the ReadMoreX widget in your layout:
#### `Read More`
```Dart
ReadMoreX(
    'This is a very long text. You can tap the "Read More" button to see more text. Suspendisse interdum consectetur libero id faucibus nisl tincidunt eget nullam. Eu nisl nunc mi ipsum faucibus vitae aliquet nec.',
    textStyle: const TextStyle(fontSize: 18),
    readMoreColor: Colors.green,
    fontWeightLabel: FontWeight.normal,
),
```
#### `Read More with Filter`
```Dart
ReadMoreX(
    'Link: https://github.com/tro1d\nAlternative link: https://flutter.dev\nEmail: readmorex@demo.com\nPhone: 0808080889',
    textStyle: const TextStyle(fontSize: 18),
    readMoreColor: Colors.green,
    filterContent: true,
    fontWeightLabel: FontWeight.normal,
    customFilter: <ReadMoreXPattern>[
        ReadMoreXPattern(
            pattern: r'github.com', // Sensitive https://
            valueChanged: (value) => value?.replaceFirst('https://github.com/', 'Github '),
            onTap: (value) {
                ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('This Link $value')),
                );
            },
        ),
        ReadMoreXPattern(
            pattern: r'https://', // Sensitive https:// , after github.com
            onTap: (value) {
                ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('This Link $value')),
                );
            },
        ),
        ReadMoreXPattern(
            pattern: r'\b[0-9]{9,}\b',
            colorChanged: Colors.red,
            onTap: (value) {
                ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('This Number $value')),
                );
            },
        ),
        ReadMoreXPattern(
            pattern: r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b',
            colorChanged: Colors.deepPurple,
            onTap: (value) {
                ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('This Email $value')),
                );
            },
        ),
    ],
),
```

Full example can be found at [`example/lib/main.dart`](https://github.com/tro1d/read_morex/blob/master/example/lib/main.dart).


CHANGELOG.md
## 1.1.0
* ✅ Fixing the bug pattern and others.
## 1.0.2
* ✅ Fixing the bug of clicking a highlight before the text is expanded.
## 1.0.1
* Update Example
## 1.0.0
* Initial release.


test/read_morex_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:read_morex/read_morex.dart';

void main() {
  testWidgets('ReadMoreX widget displays correctly', (WidgetTester tester) async {
    await tester.pumpWidget(
      const MaterialApp(
        home: ReadMoreX('Example'),
      ),
    );
    expect(find.text('Example', findRichText: true), findsOneWidget);
  });
}


lib/src/read_morex.dart
part of read_morex;

class ReadMoreX extends StatefulWidget {
  /// The text content to be displayed.
  final String content;

  /// The label to display when showing more content.
  final String? readMoreLabel;

  /// The label to display when showing less content.
  final String? showLessLabel;

  /// The font weight for the labels.
  final FontWeight? fontWeightLabel;

  /// The maximum number of lines to display before showing the "Read more" link.
  /// Default value is 3 if not specified.
  final int? maxLine;

  /// The maximum length of the text before truncation.
  /// Default value is 160 if not specified.
  final int? maxLength;

  /// The style to apply to the text.
  final TextStyle? textStyle;

  /// How the text should be aligned horizontally.
  final TextAlign? textAlign;

  /// Whether to filter the content using custom patterns.
  final bool filterContent;

  /// Custom filters to apply to the content using patterns.
  ///
  /// Example:
  /// ```dart
  /// customFilter: [
  ///   ReadMoreXPattern(
  ///     pattern: r'github.com',
  ///     valueChanged: (value) => value?.replaceFirst('https://github.com/', 'Github '),
  ///     onTap: (value) {
  ///       ScaffoldMessenger.of(context).showSnackBar(
  ///         SnackBar(content: Text('This Link $value')),
  ///       );
  ///     },
  ///   ),
  ///   ReadMoreXPattern(
  ///     pattern: r'https://', // https Pattern
  ///     onTap: (value) {
  ///       ScaffoldMessenger.of(context).showSnackBar(
  ///         SnackBar(content: Text('This Link $value')),
  ///       );
  ///     },
  ///   ),
  ///   ReadMoreXPattern(
  ///     pattern: r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', // Email Pattern
  ///     colorChanged: Colors.deepPurple,
  ///     onTap: (value) {
  ///       ScaffoldMessenger.of(context).showSnackBar(
  ///         SnackBar(content: Text('This Email $value')),
  ///       );
  ///     },
  ///   ),
  /// ],
  /// ```
  ///
  final List<ReadMoreXPattern>? customFilter;

  /// The color to use for filtered content.
  final Color? readMoreColor;

  /// The [ReadMoreX] widget is designed to display text content with the option to expand and collapse it.
  /// It allows you to set a maximum number of lines and a maximum length for the text.
  /// When the content exceeds these limits, it provides a 'Read more' link, allowing users to see
  /// the full content, and a 'Show less' link to collapse it again. You can also apply custom text filters
  /// using regular expressions to highlight or modify specific patterns within the text.
  /// This widget offers flexibility and control over how text content is displayed and filtered within your Flutter app.
  const ReadMoreX(
    this.content, {
    this.readMoreLabel,
    this.showLessLabel,
    this.fontWeightLabel,
    this.maxLine,
    this.maxLength,
    this.textStyle,
    this.textAlign,

    /// Set this to true if you want to filter the content using custom patterns.
    this.filterContent = false,

    /// Specify custom filters to apply to the content using patterns.
    this.customFilter,

    /// The color to use for filtered content.
    this.readMoreColor,
    Key? key,
  }) : super(key: key);

  @override
  State<ReadMoreX> createState() => _ReadMoreXState();
}

class _ReadMoreXState extends State<ReadMoreX> {
  bool isReadMore = false;
  String pNewline = r' @@newLine@@=> ';
  String pNewTab = r' @@newTab@@=> ';
  String pSpace = r' ';

  String content = '';
  String readMoreLabel = '';
  String showLessLabel = '';
  FontWeight fontWeightLabel = FontWeight.w600;
  int maxLine = 3;
  int maxLength = 160;
  TextStyle textStyle = const TextStyle();
  TextAlign? textAlign;
  bool filterContent = false;
  List<ReadMoreXPattern> customFilter = [];
  Color readMoreColor = const Color(0xFF2196F3);

  @override
  void initState() {
    super.initState();
    content = widget.content;
    readMoreLabel = widget.readMoreLabel ?? '...Read more';
    showLessLabel = widget.showLessLabel ?? 'Show less';
    fontWeightLabel = widget.fontWeightLabel ?? FontWeight.w600;
    maxLine = widget.maxLine ?? 3;
    maxLength = widget.maxLength ?? 160;
    textStyle = widget.textStyle ?? const TextStyle();
    textAlign = widget.textAlign;
    filterContent = widget.filterContent;
    customFilter = widget.customFilter ?? [];
    readMoreColor = widget.readMoreColor ?? const Color(0xFF2196F3);
  }

  List<InlineSpan> get textSpanChildren {
    if (filterContent) {
      return maximumWords(
        filter(
          filterPattern(
            filterLineAndLength(content),
          ).trim(),
        ),
      );
    } else {
      return maximumWords(
        [TextSpan(text: filterLineAndLength(content))],
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Text.rich(
      style: textStyle,
      textAlign: textAlign,
      TextSpan(
        children: textSpanChildren,
      ),
    );
  }

  List<InlineSpan> maximumWords(List<InlineSpan> listSpan) {
    List<InlineSpan> result = listSpan;
    final List<String> splitLine = content.split('\n');
    if (content.length > (maxLength < 1 ? 1 : maxLength) || splitLine.length >= (maxLine < 1 ? 1 : maxLine)) {
      result.add(
        WidgetSpan(
          child: GestureDetector(
            onTap: () => setState(() => isReadMore = !isReadMore),
            child: Text(
              isReadMore ? showLessLabel : readMoreLabel,
              style: textStyle.copyWith(
                color: readMoreColor,
                fontWeight: fontWeightLabel,
              ),
            ),
          ),
        ),
      );
    }
    return result;
  }

  String filterLineAndLength(String data) {
    String result = '';
    if (!isReadMore) {
      final List<String> splitLine = data.split('\n');
      final bool isMaxLength = data.length > (maxLength < 1 ? 1 : maxLength);
      final bool isMaxLine = splitLine.length >= (maxLine < 1 ? 1 : maxLine);
      if (isMaxLength) {
        result = data.substring(0, maxLength);
      } else if (isMaxLine) {
        final words = '${splitLine[0]}${splitLine[1]}${splitLine[2]}';
        result = data.substring(0, words.length);
      } else {
        result = data;
      }
    } else {
      result = data;
    }
    return result;
  }

  List<InlineSpan> filter(String data) {
    List<InlineSpan> listSpan = [];
    List<String> splitSpace = data.split(pSpace);
    for (int i = 0; i < splitSpace.length; i++) {
      final word = splitSpace[i];
      final match = RegExp(r'@@(.*?)@@=>').firstMatch(word);
      if (match == null) {
        listSpan.add(TextSpan(text: '$word '));
      } else {
        if (pNewline.contains(match.group(0)!)) listSpan.add(const TextSpan(text: '\n'));
        if (pNewTab.contains(match.group(0)!)) listSpan.add(const TextSpan(text: '\t'));
        for (ReadMoreXPattern pattern in customFilter) {
          if ('@@${pattern.pattern}@@=>'.contains(match.group(0)!)) {
            final contentWord = word.replaceFirst(match.pattern, '');
            final hasValueChanged = pattern.valueChanged != null && pattern.valueChanged!.call(contentWord)!.isNotEmpty;
            listSpan.add(
              WidgetSpan(
                child: GestureDetector(
                  onTap: () {
                    if (isReadMore) {
                      pattern.onTap?.call(contentWord);
                    } else {
                      setState(() => isReadMore = !isReadMore);
                    }
                  },
                  child: Text(
                    hasValueChanged ? '${pattern.valueChanged?.call(contentWord)} ' : '$contentWord ',
                    style: textStyle.copyWith(
                      color: pattern.colorChanged ?? readMoreColor,
                      decoration: pattern.textDecoration,
                    ),
                  ),
                ),
              ),
            );
          }
        }
      }
    }
    return listSpan;
  }

  String filterPattern(String data) {
    String result = '';
    if (customFilter.isNotEmpty) {
      String rep = data.replaceAll(RegExp(r'\n'), pNewline);
      rep = rep.replaceAll(RegExp(r'\t'), pNewTab);
      for (var i = 0; i < customFilter.length; i++) {
        if (i < 1) {
          result = _filterPattern(rep, customFilter[i]);
        } else {
          result = _filterPattern(result, customFilter[i]);
        }
      }
    }
    return result;
  }

  String _filterPattern(String data, ReadMoreXPattern patterns) {
    final splitSpace = data.split(pSpace);
    final res = List.generate(
      splitSpace.length,
      (index) {
        final word = splitSpace[index];
        if (!word.contains(RegExp(patterns.pattern))) {
          return '$word$pSpace';
        } else if (word.contains(RegExp(pNewline)) || word.contains(RegExp(pNewTab))) {
          return '$word$pSpace';
        } else {
          if (word.contains(RegExp(r'@@(.*?)@@=>'))) {
            return '$word$pSpace';
          } else {
            return '@@${patterns.pattern}@@=>$word$pSpace';
          }
        }
      },
    );
    return res.join();
  }
}


example/lib/post_card.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:read_morex/read_morex.dart';

class PostCard extends StatelessWidget {
  const PostCard({Key? key, required this.userName, required this.content}) : super(key: key);

  final String userName;
  final String content;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.green.shade50,
      padding: const EdgeInsets.symmetric(horizontal: 16.0),
      margin: const EdgeInsets.only(bottom: 4),
      child: Column(
        children: <Widget>[
          Row(
            children: <Widget>[
              ClipRRect(
                borderRadius: BorderRadius.circular(26),
                child: const Icon(CupertinoIcons.person_alt),
              ),
              Expanded(
                child: Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 6.0),
                  child: Text(userName, style: const TextStyle(fontWeight: FontWeight.w600)),
                ),
              ),
              const Padding(
                padding: EdgeInsets.symmetric(horizontal: 6.0),
                child: Text('4 minutes ago', style: TextStyle(fontSize: 10, color: Colors.black45)),
              ),
              const Icon(Icons.more_horiz_rounded)
            ],
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 8.0),
            child: Column(
              children: <Widget>[
                Align(
                  alignment: Alignment.centerLeft,
                  child: ReadMoreX(
                    content,
                    textStyle: const TextStyle(fontSize: 18),
                    readMoreColor: Colors.green,
                    filterContent: true,
                    fontWeightLabel: FontWeight.normal,
                    customFilter: <ReadMoreXPattern>[
                      ReadMoreXPattern(
                        pattern: r'github.com',
                        valueChanged: (value) => value?.replaceFirst('https://github.com/', 'Github '),
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Number $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'https://',
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Link $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'\b[0-9]{9,}\b',
                        colorChanged: Colors.red,
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Number $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b',
                        colorChanged: Colors.deepPurple,
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Email $value')),
                          );
                        },
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
          const Align(
            alignment: Alignment.centerLeft,
            child: Text.rich(
              TextSpan(
                children: <InlineSpan>[
                  TextSpan(text: 'Liked by '),
                  TextSpan(text: 'You ', style: TextStyle(fontWeight: FontWeight.w600)),
                  TextSpan(text: 'and '),
                  TextSpan(
                    text: '99 Others',
                    style: TextStyle(fontWeight: FontWeight.w600),
                  ),
                ],
              ),
            ),
          ),
          const Row(
            children: <Widget>[
              Icon(CupertinoIcons.ellipses_bubble, size: 20),
              SizedBox(width: 8),
              Icon(
                CupertinoIcons.heart_fill,
                color: Colors.green,
                size: 22,
              ),
              SizedBox(width: 8),
              Icon(Icons.send_rounded, size: 20),
              Expanded(
                child: Text(
                  '63 Comments',
                  textAlign: TextAlign.end,
                ),
              ),
            ],
          ),
          const Divider(),
        ],
      ),
    );
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "read_morex_example");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "read_morex_example");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"read_morex_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/src/read_morex_pattern.dart
part of read_morex;

class ReadMoreXPattern {
  /// The regular expression pattern to be used for text filtering.
  ///
  /// Example:
  ///```Dart
  /// pattern: r'https://',
  ///```
  final String pattern;

  /// A function to transform the filtered value.
  final String? Function(String?)? valueChanged;

  /// The color to apply to the filtered text.
  final Color? colorChanged;

  /// A callback function that is triggered when the filtered text is tapped.
  final ValueChanged<String?>? onTap;

  /// A linear decoration to draw near the filter content.
  final TextDecoration? textDecoration;

  ReadMoreXPattern({
    required this.pattern,

    /// A function to modify the filtered value based on the pattern.
    this.valueChanged,

    /// The color to apply to the filtered text.
    this.colorChanged,

    /// A callback function to handle taps on the filtered text.
    this.onTap,

    /// Additional text decoration to be applied to the filtered text.
    this.textDecoration,
  });
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:read_morex/read_morex.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(colorSchemeSeed: Colors.green),
      home: Scaffold(
        appBar: AppBar(title: const Text('Read MoreX Example')),
        body: const Padding(
          padding: EdgeInsets.symmetric(vertical: 16.0),
          child: SingleChildScrollView(
            child: Column(
              children: [
                PostCard(
                  userName: 'Read More',
                  content:
                      'This is a very long text. You can tap the "Read More" button to see more text. Suspendisse interdum consectetur libero id faucibus nisl tincidunt eget nullam. Eu nisl nunc mi ipsum faucibus vitae aliquet nec. Suspendisse interdum consectetur libero id faucibus nisl tincidunt eget nullam. Eu nisl nunc mi ipsum faucibus vitae aliquet nec.',
                ),
                PostCard(
                  userName: 'Read More with Filter',
                  content: 'Link: https://github.com/tro1d\nAlternative link: https://flutter.dev\nEmail: readmorex@demo.com\nPhone: 0808080889',
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class PostCard extends StatelessWidget {
  const PostCard({Key? key, required this.userName, required this.content}) : super(key: key);

  final String userName;
  final String content;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.green.shade50,
      padding: const EdgeInsets.symmetric(horizontal: 16.0),
      margin: const EdgeInsets.only(bottom: 4),
      child: Column(
        children: <Widget>[
          Row(
            children: <Widget>[
              ClipRRect(
                borderRadius: BorderRadius.circular(26),
                child: const Icon(CupertinoIcons.person_alt),
              ),
              Expanded(
                child: Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 6.0),
                  child: Text(userName, style: const TextStyle(fontWeight: FontWeight.w600)),
                ),
              ),
              const Padding(
                padding: EdgeInsets.symmetric(horizontal: 6.0),
                child: Text('4 minutes ago', style: TextStyle(fontSize: 10, color: Colors.black45)),
              ),
              const Icon(Icons.more_horiz_rounded)
            ],
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 8.0),
            child: Column(
              children: <Widget>[
                Align(
                  alignment: Alignment.centerLeft,
                  child: ReadMoreX(
                    content,
                    textStyle: const TextStyle(fontSize: 18),
                    readMoreColor: Colors.green,
                    filterContent: true,
                    fontWeightLabel: FontWeight.normal,
                    customFilter: <ReadMoreXPattern>[
                      ReadMoreXPattern(
                        pattern: r'github.com',
                        valueChanged: (value) => value?.replaceFirst('https://github.com/', 'Github '),
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Number $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'https://',
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Link $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'\b[0-9]{9,}\b',
                        colorChanged: Colors.red,
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Number $value')),
                          );
                        },
                      ),
                      ReadMoreXPattern(
                        pattern: r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b',
                        colorChanged: Colors.deepPurple,
                        onTap: (value) {
                          ScaffoldMessenger.of(context).showSnackBar(
                            SnackBar(content: Text('This Email $value')),
                          );
                        },
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
          const Align(
            alignment: Alignment.centerLeft,
            child: Text.rich(
              TextSpan(
                children: <InlineSpan>[
                  TextSpan(text: 'Liked by '),
                  TextSpan(text: 'You ', style: TextStyle(fontWeight: FontWeight.w600)),
                  TextSpan(text: 'and '),
                  TextSpan(
                    text: '99 Others',
                    style: TextStyle(fontWeight: FontWeight.w600),
                  ),
                ],
              ),
            ),
          ),
          const Row(
            children: <Widget>[
              Icon(CupertinoIcons.ellipses_bubble, size: 20),
              SizedBox(width: 8),
              Icon(
                CupertinoIcons.heart_fill,
                color: Colors.green,
                size: 22,
              ),
              SizedBox(width: 8),
              Icon(Icons.send_rounded, size: 20),
              Expanded(
                child: Text(
                  '63 Comments',
                  textAlign: TextAlign.end,
                ),
              ),
            ],
          ),
          const Divider(),
        ],
      ),
    );
  }
}
